piece-retriever/lib/piece.js

```diff
@@ -0,0 +1,48 @@
+import * as Hasher from '@web3-storage/data-segment/multihash'
+import * as Raw from 'multiformats/codecs/raw'
+import * as Link from 'multiformats/link'
+
+/** @typedef {import('@web3-storage/data-segment').PieceLink} PieceCID */
+
+/**
+ * @returns {{
+ *   stream: TransformStream<Uint8Array, Uint8Array>
+ *   getPieceCID: () => PieceCID | null
+ * }}
+ */
+export function createPieceCIDStream() {
+  const hasher = Hasher.create()
+  let finished = false
+  /** @type {PieceCID | null} */
+  let pieceCid = null
+
+  /** @type {TransformStream<Uint8Array, Uint8Array>} */
+  const stream = new TransformStream({
+    /**
+     * @param {Uint8Array} chunk
+     * @param {TransformStreamDefaultController<Uint8Array>} controller
+     */
+    transform(chunk, controller) {
+      hasher.write(chunk)
+      controller.enqueue(chunk)
+    },
+
+    flush() {
+      // Calculate final PieceCID when stream ends
+      const digest = hasher.digest()
```

Comment:
I find it suspicious that the code calculating the hash is fully synchronous and there is no `await` anywhere. Are we missing an `await` statement somewhere?

Node.js crypto API and WebCrypto use async operations for calculating the hashes in order to move the CPU-expensive operations off the main JavaScript thread.

I am not familiar with `@web3-storage/data-segment/multihash`. If it works synchronously, then please ignore this comment.